### PR TITLE
Fix button focus

### PR DESCRIPTION
**Pull request by faridomarAf**
### Description
While we are using a Link-component for both Button & Link purposes, there should be a logic when it use as a Link it should acts as a Link, and when it use as Button it should as Button.

### 1: Enhancements to Button Component::
- **Role Handling:** Added logic to differentiate between button and link usage based on the presence of the href prop.
- **Keyboard Accessibility:** Implemented onKeyDown to handle Enter and Space keys, ensuring proper behavior for keyboard users.
- **Disabled State Management:** Updated to respect disabled prop:
   - Prevented clicks and keyboard interaction when button disabled.
   - Removed from the **tab** order using tabIndex when disabled.
   - Applied a disabled style conditionally.
- **Cleaner Role Logic**: Dynamically set `role` to `button` only when used as a button.
- **Improved Prop Handling:** Ensured `onClick` and `tabIndex` behave appropriately based on component usage.

### 2: Enhancements to CodeBox Component:
- **Dynamic Button State:**
   - Added a `useState` hook to manage the disabled state of the `Copy` button.
   - Used `useEffect` to dynamically update the button's `disabled` state based on the presence of content in the `<pre> `element.
- **Keyboard Accessibility:**
  - Ensured proper tab behavior with `tabIndex` on the` <pre>` element for improved accessibility.
- **Improved User Experience:**
  - Prevented users from interacting with the `Copy` button if no content is available to copy.

## Validation
https://www.loom.com/share/6f237ae46e7945f48650c1e710399e6f?sid=f8f6c2fc-faca-46e8-901a-021c38f091f9

## Related Issues

https://github.com/nodejs/nodejs.org/issues/7247

### Check List

- [ ] I have read the [Contributing Guidelines](https://github.com/nodejs/nodejs.org/blob/main/CONTRIBUTING.md) and made commit messages that follow the guideline.
- [x] I have run `npm run format` to ensure the code follows the style guide.
- [x] I have run `npm run test` to check if all tests are passing.
- [x] I have run `npx turbo build` to check if the website builds without errors.
- [ ] I've covered new added functionality with unit tests if necessary.

### CODE REVIEW

**Review by bjohansebas — COMMENTED** [review]
LGTM

File: apps/site/components/Containers/MetaBar/index.module.css
Comment (mikeesto) [inline apps/site/components/Containers/MetaBar/index.module.css:1]: Can you please remove these changes? They don't seem to be related. Thanks

Comment (faridomarAf) [reply apps/site/components/Containers/MetaBar/index.module.css:1]: its done

File: apps/site/components/Common/CodeBox/index.tsx [hunk 1]
```diff
@@ -114,7 +114,12 @@ const CodeBox: FC<PropsWithChildren<CodeBoxProps>> = ({
           <span className={styles.language}>{language}</span>
 
           {showCopyButton && (
-            <Button kind="neutral" className={styles.action} onClick={onCopy}>
+            <Button
+              kind="neutral"
+              className={styles.action}
+              onClick={onCopy}
+              href="#"
```
Comment (AugustinMauroy) [inline apps/site/components/Common/CodeBox/index.tsx]: I've never heard of this being a question of accessibility. Do you have a WAI document that shows this?

**Review by mikeesto — COMMENTED** [review]
I don't understand why this component isn't a native `<button>` (there may well be a good reason). I think this change is an improvement because now it is focusable, but the default behaviour of `href="#"` is to jump the page to the top so that's not ideal either

**Review by ovflowd — COMMENTED** [review]
I don't think this is an appropriate fix, we should be using tabindex and other properties.

**Comment by ovflowd** [comment]
> I don't understand why this component isn't a native `<button>` (there may well be a good reason). I think this change is an improvement because now it is focusable, but the default behaviour of `href="#"` is to jump the page to the top so that's not ideal either

There are numerous reasons, mostly because all our "buttons" are links, this one is deceptively the only one that is not a button. Per specification a `a` element with `role` button, for all intends and purposes _is a button_; So that's not the issue.

**Comment by faridomarAf** [comment]
yeah guys this is not the proper way, sorry for late replying, you are right 
let me check the components which this component is used in those because there are some functionality for "Enter key" and "Space key" that should be considered and also we should check is there any necessity for the "href" or no, whit checking all these, I would come whit a better solution for that which can be logically satisfying 😊

**Comment by faridomarAf** [comment]
hey guys @mikeesto @AugustinMauroy @ovflowd @bjohansebas could you please review my code, if there is any suggestion PLZ  😊

File: apps/site/components/Common/Button/index.tsx [hunk 2]
```diff
@@ -17,17 +17,40 @@ const Button: FC<ButtonProps> = ({
...
-  >
-    {children}
-  </Link>
-);
+}) => {
+  // Check if the component is being used as a button (no href provided)
+  const isButton = !href;
+
+  // Handle keydown events for keyboard accessibility
+  const handleKeyDown = (e: React.KeyboardEvent<HTMLAnchorElement>) => {
```
Comment (AugustinMauroy) [inline apps/site/components/Common/Button/index.tsx]: on import part of this file you need to do something like this
```tsx
import type { KeyboardEvent } from 'react';
```

File: apps/site/components/Common/Button/index.tsx [hunk 3]
```diff
@@ -17,17 +17,40 @@ const Button: FC<ButtonProps> = ({
...
+    if (disabled) return;
+
+    if (e.key === 'Enter' || e.key === ' ') {
+      e.preventDefault();
+      if (onClick) {
+        onClick(e as unknown as React.MouseEvent<HTMLAnchorElement>);
```
Comment (AugustinMauroy) [inline apps/site/components/Common/Button/index.tsx]: same here

File: apps/site/components/Common/Button/index.tsx [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import classNames from 'classnames';
 import type { FC, AnchorHTMLAttributes } from 'react';
```
Comment (mikeesto) [inline apps/site/components/Common/Button/index.tsx]:
```suggestion
import type { FC, AnchorHTMLAttributes, KeyboardEvent, MouseEvent } from 'react';
```

File: apps/site/components/Common/Button/index.tsx [hunk 5]
```diff
@@ -1,5 +1,6 @@
...
+import type { KeyboardEvent, MouseEvent } from 'react';
```
Comment (mikeesto) [inline apps/site/components/Common/Button/index.tsx]:
```suggestion

```

**Review by mikeesto — COMMENTED** [review]
Button focus seems good to me 👍

**Review by bjohansebas — COMMENTED** [review]
Everything seems to be working fine, LGTM

File: apps/site/components/Common/Button/index.tsx (hunk 5 above)
Comment (faridomarAf) [reply apps/site/components/Common/Button/index.tsx]: yeah I refactored that, and also pushed I don't know why it dos not pushed 🥸

Comment (faridomarAf) [reply apps/site/components/Common/Button/index.tsx]: now its done 😊

File: apps/site/components/Common/Button/index.tsx [hunk 6]
```diff
@@ -17,17 +22,40 @@ const Button: FC<ButtonProps> = ({
...
+
+  return (
+    <Link
+      role={isButton ? 'button' : undefined} // Set role to 'button' only when used as a button
+      href={disabled ? undefined : href} // Disable the link when disabled
+      aria-disabled={disabled}
+      className={classNames(styles.button, styles[kind], className, {
+        [styles.disabled]: disabled, // Add disabled style when appropriate
+      })}
+      tabIndex={isButton ? (disabled ? -1 : 0) : undefined} // Remove from tab order if disabled for buttons
```
Comment (bmuenzenmeyer) [inline apps/site/components/Common/Button/index.tsx]: should we be instead (or additionally) using `<button disabled=true />` here - it supports skipping the tab order

https://codepen.io/bmuenzenmeyer/pen/KwPgNGe/9d52f7e51fbf7a037b499caed800da9e

Comment (bmuenzenmeyer) [reply apps/site/components/Common/Button/index.tsx]: nevermind, I see it passing through from props

Comment (faridomarAf) [reply apps/site/components/Common/Button/index.tsx]: the idea is when the button is disabled it should not be focusable

**Review by bmuenzenmeyer — APPROVED** [review]

**Comment by bmuenzenmeyer** [comment]
@faridomarAf  please resolve the CI failures if you have time

**Comment by ovflowd** [comment]
Apologies for not checking in here, have been busy with personal matters 🙏

**Comment by faridomarAf** [comment]
hi @bmuenzenmeyer , sorry for late response, I think the changes which I made in Button.tsx file, it caused the error, so a I checked the original code that is working, I would do my best to solve the issue

File: apps/site/components/Common/Button/index.tsx
Comment (ovflowd) [inline apps/site/components/Common/Button/index.tsx:1]: Please add "use client" at the top, since you are adding interactivity, unfortunately this needs to be a client component.

File: apps/site/components/Common/Button/index.tsx [hunk 7]
```diff
@@ -17,17 +22,46 @@ const Button: FC<ButtonProps> = ({
...
-    className={classNames(styles.button, styles[kind], className)}
-    {...props}
-  >
-    {children}
-  </Link>
-);
+}) => {
+  //to handle the keyboard interactions, specifically for Enter and Space keys
+  const handleKeyDown = (e: KeyboardEvent<HTMLAnchorElement>) => {
+    if (disabled) return;
```
Comment (ovflowd) [inline apps/site/components/Common/Button/index.tsx]: Add brackets please. And shouldn't there be a preventDefault?

File: apps/site/components/Common/Button/index.tsx [hunk 8]
```diff
@@ -17,17 +22,46 @@ const Button: FC<ButtonProps> = ({
...
+    }
+  };
+
+  // to manage mouse click events for the component, providing behavior consistent with the disabled state
+  const handleClick = (e: MouseEvent<HTMLAnchorElement>) => {
+    if (disabled) {
+      e.preventDefault();
+      return;
+    }
+    if (onClick) {
```
Comment (ovflowd) [inline apps/site/components/Common/Button/index.tsx]:
```suggestion

    if (typeof onClick === 'function') {
```

File: apps/site/components/Common/Button/index.tsx [hunk 9]
```diff
@@ -17,17 +22,46 @@ const Button: FC<ButtonProps> = ({
...
+  };
+
+  return (
+    <Link
+      role="button"
+      href={disabled ? undefined : href}
+      aria-disabled={disabled}
+      className={classNames(styles.button, styles[kind], className)}
+      tabIndex={disabled ? -1 : 0} // Ensure focusable if not disabled
+      onClick={handleClick}
```
Comment (ovflowd) [inline apps/site/components/Common/Button/index.tsx]: IMO instead of disabled on the function call, you should do:

```suggestion
      onClick={disabled || onClick}
```

Or something similar

Comment (ovflowd) [reply apps/site/components/Common/Button/index.tsx]: On both the onClick and the onKeyDown probably. Not sure if you need to override the default behaviour, but you could also do:

```jsx
onClick={disabled ? preventDefaultFn : onClickHandler}
```

File: apps/site/components/Common/Button/index.tsx [hunk 10]
```diff
@@ -17,17 +22,46 @@ const Button: FC<ButtonProps> = ({
...
+    if (e.key === 'Enter' || e.key === ' ') {
+      e.preventDefault();
+      if (onClick) {
+        onClick(e as unknown as MouseEvent<HTMLAnchorElement>);
+      }
+    }
+  };
+
+  // to manage mouse click events for the component, providing behavior consistent with the disabled state
+  const handleClick = (e: MouseEvent<HTMLAnchorElement>) => {
```
Comment (ovflowd) [inline apps/site/components/Common/Button/index.tsx]: I feel that calling it onClickHandler sounds more concise with the naming of the event itself

File: apps/site/components/Common/Button/index.tsx (hunk 9 above)
Comment (faridomarAf) [reply apps/site/components/Common/Button/index.tsx]: I think keeping the disabled logic in the handleClick function for clarity, maintainability, and accessibility, as it ensures consistent behavior for both mouse and keyboard interactions. onClick={disabled || onClick} risks unintended behavior and lacks flexibility for future enhancements

File: apps/site/components/Common/Button/index.tsx [hunk 11]
```diff
@@ -17,17 +24,49 @@ const Button: FC<ButtonProps> = ({
...
-    aria-disabled={disabled}
-    className={classNames(styles.button, styles[kind], className)}
-    {...props}
-  >
-    {children}
-  </Link>
-);
+}) => {
+  //to handle the keyboard interactions, specifically for Enter and Space keys
+  const handleKeyDown = (e: KeyboardEvent<HTMLAnchorElement>) => {
```
Comment (ovflowd) [inline apps/site/components/Common/Button/index.tsx]: onKeyDownHandler :)

File: apps/site/components/Common/Button/index.tsx [hunk 12]
```diff
@@ -17,17 +24,49 @@ const Button: FC<ButtonProps> = ({
...
+
+    if (e.key === 'Enter' || e.key === ' ') {
+      e.preventDefault();
+      if (typeof onClick === 'function') {
+        onClick(e as unknown as MouseEvent<HTMLAnchorElement>);
+      }
+    }
+  };
+
+  // to manage mouse click events for the component, providing behavior consistent with the disabled state
```
Comment (ovflowd) [inline apps/site/components/Common/Button/index.tsx]: I dont think these inline comments bring any value, it is quite evident already what these functions do :)

File: apps/site/components/Common/Button/index.tsx (hunk 8 above)
Comment (ovflowd) [reply apps/site/components/Common/Button/index.tsx]: @faridomarAf 👀

**Review by ovflowd — APPROVED** [review]
LGTM!

**Comment by ovflowd** [comment]
Thanks for the first time contribution! Excited to get this merged ✨

**Comment by faridomarAf** [comment]
hey guys thank you so much for your follow up, actually it was my fist time to contribute in an open source project. all your comments an code views values much more for me 
thanks 🙏
